### ISSUE TITLE
Message size calculations discrepancy

### ISSUE THREAD

**Issue by malimohub** [co-go/sqs-extended-client-go#104]
Seems to be an issue in sqs extended client msg size calculations, where the value is about .3% less than msg size calculated on AWS servers. Thought by including missing DataType msg attr in message size cals (https://github.com/co-go/sqs-extended-client-go/pull/103) would resolve issue, however seems other message properties are being assessed in size calculations on AWS servers.

**Comment by co-go** [comment]
@malimohub Can you provide a screenshot/link to where AWS is providing the message sizes? Are you referring to the CloudWatch metrics for the message size in your queue?

**Comment by malimohub** [comment]
@co-go we were seeing errors from AWS, and logged out the AWS error containing the size they calculated and the limit. "AWS.SimpleQueueService.BatchRequestTooLong: Batch requests cannot be longer than 262144 bytes. You have sent 262953 bytes"

**Comment by co-go** [comment]
Ah, it looks like I'll need to take a closer look into what the calculations are for the **Batch** messages. The individual messages could be correctly calculated, but this error message for sure means the entire batch size is not calculated correctly as of now. Thanks for following up!

**Comment by malimohub** [comment]
> Ah, it looks like I'll need to take a closer look into what the calculations are for the **Batch** messages. The individual messages could be correctly calculated, but this error message for sure means the entire batch size is not calculated correctly as of now. Thanks for following up!

sweet, let me know if i can help in anyway. I appreciate you taking a look

**Comment by co-go** [comment]
So without knowing the exact sizing of your payloads, it is hard to say **exactly** how you exceeded the message size, but let me try to summarize some of the gotchas I found when looking into this.

Firstly, just to clarify and set context. We are often talking about size limits in this package, but specifically with batch messages, there is an important difference between the *message* size limit and the *batch* size limit, even though AWS defaults these both to 256 KB. 

- **Message** size limit refers to the actual memory needed to store the **raw message and its attributes**. From my testing with multiple different message sizes and attribute types, I was unable to find a bug in how this is currently calculated.
- **Batch** size limit refers to the size of the **entire JSON payload being sent to AWS** when sending the API call. This can actually add a decent bit more overhead with all the extra attributes needed for the JSON that might not be reflected in the actual messages.

Here is an example of a payload where I am utilizing `BatchSendMessage` with a single message and an attribute:
```json
{"Entries":[{"Id":"some-msg","MessageAttributes":{"attr1":{"DataType":"String","StringValue":"str"}},"MessageBody":"some msg"}],"QueueUrl":"https://sqs.us-east-1.amazonaws.com/520156417308/sqsec-integration-queue"}
```
This entire payload is **214 bytes** whereas the underlying message to be stored is only **22 bytes**.

Now, when sending very large payloads, the "overhead" relative to the actual data being sent will be much less, but I think this is an important callout to make as it could impact some edge cases unless we account for this when sending batch payloads. This is problem 1 that should be addressed.

Next, moving onto the bigger, and more likely, reason why I think you are running into this issue.

Currently we calculate the size of each message, and if adding that to our batch causes the size to exceed the limit, then the message is replaced with the pointer. Hypothetically, if you were to add one message that was 255 KB, just below the default limit and would not be replaced with a pointer, then proceed to add another 9 large messages, even though their message contents will be replaced with pointers before being added to the batch, the pointer bodies + attributes themselves would likely be enough to cause the entire batch to still exceed the batch limit. The solution to this would be to replace the first message payload with an extended pointer as well so we don't exceed the batch limit.

Relatively, this second issue is more likely to cause bugs than the first, but both of these should be accounted for in our `BatchSendMessage` in order to make sending many large payloads more consistent.

Thanks for your patience- I'll start working on a fix for these, let me know if you have any other thoughts!

**Comment by co-go** [comment]
After implementing some more tests to make sure I was correctly solving for this problem, I realized I made a mistake in my comment above. **Batch** size limit is actually much simpler and only refers to the sum of all messages and their attributes. What this means is actually the first problem (overhead with encoding) is actually a non-issue and can be disregarded. However, the second problem is still very much a bug with the current implementation.
